lib/stm32/common/fdcan.c

```diff
@@ -0,0 +1,763 @@
+{
+	const struct fdcan_message_ram *ram = fdcan_get_msgram_addr(canport);
+
+	const struct fdcan_rx_fifo_element *fifo;
+
+	unsigned pending_frames, get_index, dlc, len;
+
+	if (fdcan_get_fill_rxfifo(canport, fifo_id, &get_index, &pending_frames) != FDCAN_E_OK) {
+		return FDCAN_E_INVALID;
+	}
+
+	if (fifo_id == 0) {
+		fifo = ram->rx_fifo[0];
+	} else  {
+		/* No need to check if fifo_id == 1 */
+		fifo = ram->rx_fifo[1];
+	}
+
+	if (pending_frames == 0) {
+		return FDCAN_E_NOTAVAIL;
+	}
+
+	dlc = (fifo[get_index].filt_fmt_dlc_ts & FDCAN_FIFO_DLC_MASK)
+		>> FDCAN_FIFO_DLC_SHIFT;
+
+	len = fdcan_dlc_to_length(dlc);
+
+	*length = len;
+	if ((fifo[get_index].identifier_flags & FDCAN_FIFO_XTD) == FDCAN_FIFO_XTD) {
+		*ext = true;
+		*id = (fifo[get_index].identifier_flags & FDCAN_FIFO_EID_MASK)
+			>> FDCAN_FIFO_EID_SHIFT;
+	} else {
+		*ext = false;
+		*id = (fifo[get_index].identifier_flags & FDCAN_FIFO_SID_MASK)
+			>> FDCAN_FIFO_SID_SHIFT;
+	}
+
+	if (timestamp) {
+		*timestamp = (uint16_t) (fifo[get_index].filt_fmt_dlc_ts & FDCAN_FIFO_RXTS_MASK)
+			>> FDCAN_FIFO_RXTS_SHIFT;
+	}
+
+	if (fmi) {
+		*fmi = (uint8_t) (fifo[get_index].filt_fmt_dlc_ts & FDCAN_FIFO_MM_MASK)
+			>> FDCAN_FIFO_MM_SHIFT;
+	}
+
+	if (rtr) {
+		*rtr = ((fifo[get_index].identifier_flags & FDCAN_FIFO_RTR) == FDCAN_FIFO_RTR);
+	}
+
+	for (unsigned int q = 0; q < len; q += 4) {
+		*((uint32_t *) &data[q]) = fifo[get_index].data[q / 4];
+	}
+
+	if (release) {
+		if (fifo_id == 0) {
+			FDCAN_RXF0A(canport) |= (get_index & FDCAN_RXF0A_R0AI_MASK)
+				<< FDCAN_RXF0A_R0AI_SHIFT;
+		} else {
+			FDCAN_RXF1A(canport) |= (get_index & FDCAN_RXF1A_R1AI_MASK)
+				<< FDCAN_RXF1A_R1AI_SHIFT;
+		}
+	}
+
+	return FDCAN_E_OK;
+}
+
+/** Release receive oldest FIFO entry.
+ * This function will mask oldest entry in FIFO as released making
+ * space for another received frame. This function can be used if
+ * fdcan_receive was called using release = false. If used in other
+ * case, then messages can get lost.
+ *
+ * @param [in] canport FDCAN block base address
+ * @param [in] fifo_id ID of FIFO where release should be performed (0 or 1)
+ * @returns Operation error status. @ref fdcan_error
+ */
+int fdcan_fifo_release(uint32_t canport, uint8_t fifo_id)
+{
+	unsigned pending_frames, get_index;
+
+	if (fifo_id == 0) {
+		get_index = (FDCAN_RXF0S(canport) & FDCAN_RXF0S_F0GI_MASK)
+			>> FDCAN_RXF0S_F0GI_SHIFT;
+		pending_frames = (FDCAN_RXF0S(canport) & FDCAN_RXF0S_F0FL_MASK)
+			>> FDCAN_RXF0S_F0FL_SHIFT;
+
+		if (pending_frames) {
+			FDCAN_RXF0A(canport) |= (get_index & FDCAN_RXF0A_R0AI_MASK)
+				<< FDCAN_RXF0A_R0AI_SHIFT;
+		}
+	} else if (fifo_id == 1) {
+		get_index = (FDCAN_RXF1S(canport) & FDCAN_RXF1S_F1GI_MASK)
+					>> FDCAN_RXF1S_F1GI_SHIFT;
+		pending_frames = (FDCAN_RXF1S(canport) & FDCAN_RXF1S_F1FL_MASK)
+			>> FDCAN_RXF1S_F1FL_SHIFT;
+
+		if (pending_frames) {
+			FDCAN_RXF1A(canport) |= (get_index & FDCAN_RXF1A_R1AI_MASK)
+				<< FDCAN_RXF1A_R1AI_SHIFT;
+		}
```

Comment:
like this section could just become 
```
get_index = FDCAN_RXFIFOS(canport, fifoid) & ___fifo_GI_mask >> __fifo_GI_SHIFT
pending = same...
FDCAN_FIFOA(can, fifoid) |= get_index && mask << shift....
```
and then just return void, and drop teh fifo check?